rust_src/src/multibyte.rs

```diff
@@ -312,6 +313,59 @@ impl LispObject {
     }
 }
 
+#[derive(Copy, Clone, PartialEq)]
+pub struct LispSymbolOrString(LispObject);
+
+impl LispSymbolOrString {
+    pub fn is_string(self) -> bool {
+        self.0.is_string()
+    }
+
+    pub fn is_symbol(self) -> bool {
+        self.0.is_symbol()
+    }
+}
+
+impl From<LispSymbolOrString> for LispObject {
+    fn from(s: LispSymbolOrString) -> Self {
+        s.0
+    }
+}
+
+impl From<LispSymbolOrString> for LispStringRef {
+    fn from(s: LispSymbolOrString) -> Self {
+        match s.0.as_symbol() {
+            Some(symbol) => symbol.symbol_name().into(),
+            None => s.0.into(),
+        }
+    }
+}
+
+impl From<LispStringRef> for LispSymbolOrString {
```

Comment:
Another option is to make `LispSymbolOrString` an enum containing `String(LispStringRef)` and `Symbol(LispSymbolRef)`. It is not clear which option is preferable. Something to experiment with.